src/analysis/locally_convex/with_seminorms.lean

```diff
@@ -289,6 +289,78 @@ begin
+  split,
+  { rintros ⟨V, V_basis, hVU⟩,
+    rcases p.basis_sets_iff.mp V_basis with ⟨_, _, r_pos, V_ball⟩,
+    simp_rw [V_ball, seminorm.singleton_add_ball _, add_zero] at hVU,
+    exact ⟨_, _, r_pos, hVU⟩ },
+  { rintros ⟨s, r, r_pos, V_sub_U⟩,
+    refine ⟨(s.sup p).ball 0 r, p.basis_sets_iff.mpr ⟨_, _, r_pos, rfl⟩, _⟩,
+    simp_rw [seminorm.singleton_add_ball _, add_zero],
+    exact V_sub_U }
+end
+
+/-- The open sets of a space whose topology is induced by a family of seminorms
+are exactly the sets which contain seminorm balls around all of their points.-/
+lemma with_seminorms.is_open_iff_mem_balls (hp : with_seminorms p) (U : set E) :
+  is_open U ↔ ∀ (x ∈ U), ∃ (s : finset ι) (r > 0), (s.sup p).ball x r ⊆ U :=
+by simp_rw [←with_seminorms.mem_nhds_iff hp _ U, is_open_iff_mem_nhds]
+
+/-- A separating family of seminorms induces a T₂ topology. -/
+lemma with_seminorms.t2_of_separating (hp : with_seminorms p) (h : ∀ x, (x ≠ 0) → ∃ i, p i x ≠ 0) :
+  t2_space E :=
+begin
+  rw t2_space_iff_nhds,
+  intros x y hxy,
+  cases h (x - y) (sub_ne_zero.mpr hxy) with i pi_nonzero,
+  let r := p i (x - y),
+  have r_div_2_pos : p i (x - y) / 2 > 0 := by positivity,
+  have mem_nhds : ∀ x, (p i).ball x (r/2) ∈ nhds x :=
+  λ x, (with_seminorms.mem_nhds_iff hp _ _).mpr ⟨_, _, r_div_2_pos, by rw finset.sup_singleton⟩,
+  use [(p i).ball x (r/2) , mem_nhds x, (p i).ball y (r/2), mem_nhds y],
+  intros W W_sub_xball W_sub_yball,
+  by_contra W_nonemp,
+  rw [le_bot_iff, bot_eq_empty, ←ne.def, ←nonempty_iff_ne_empty] at W_nonemp,
+  cases W_nonemp with z z_in_W,
+  have : r < r,
+  { calc
+    r   = p i (x - y)               : rfl
+    ... = p i ((x - z) + (z - y))   : by rw ←sub_add_sub_cancel
+    ... ≤ p i (x - z) + p i (z - y) : seminorm.add_le' _ _ _
+    ... = p i (z - x) + p i (z - y) : by rw [←map_neg_eq_map, neg_sub]
+    ... < r/2 + r/2                 : add_lt_add (W_sub_xball z_in_W) (W_sub_yball z_in_W)
+    ... = r                         : add_halves _ },
+  rwa lt_self_iff_false r at this
+end
+
+/-- A family of seminorms inducing a T₂ topology is separating. -/
+lemma with_seminorms.separating_of_t2 [t2_space E] (hp : with_seminorms p) (x : E) (hx : x ≠ 0) :
+  ∃ i, p i x ≠ 0 :=
+begin
+  rcases t2_separation hx with ⟨U, V, hU, hV, hxU, h0V, UV_disj⟩,
+  rcases (with_seminorms.is_open_iff_mem_balls hp _).mp hV 0 h0V with ⟨s, r, r_pos, ball_V⟩,
+  have hpx_nezero : (s.sup p) x ≠ 0,
+  { intro hsup_px,
+    refine (set.singleton_nonempty x).not_subset_empty (UV_disj _ _);
+      rw [set.le_eq_subset, set.singleton_subset_iff],
+    { exact hxU },
+    { refine ball_V _,
+      rwa [seminorm.mem_ball, sub_zero, hsup_px] } },
+  rw seminorm.finset_sup_apply at hpx_nezero,
+  have := hpx_nezero.symm.lt_of_le,
+  simp only [nnreal.zero_le_coe, nnreal.coe_pos,
+    finset.lt_sup_iff, exists_prop, forall_true_left] at this,
+  rcases this with ⟨i, his, hpix_pos⟩,
+  use ⟨i, ne_of_gt hpix_pos⟩
+end
```

Comment:
Ok so for this one I have a more mysterious proposition:
```suggestion
begin
  have := ((t1_space_tfae E).out 0 9).mp infer_instance,
  by_contra' h,
  refine hx (this _),
  rw hp.has_basis_zero_ball.specializes_iff,
  rintros ⟨s, r⟩ (hr : 0 < r),
  simp only [ball_finset_sup_eq_Inter _ _ _ hr, mem_Inter₂, mem_ball_zero, h, hr, forall_true_iff],
end
```
The key fact here is that you only need `T1`, and `T1` has a nice characterization in terms of the squiggly arrow `⤳` (see `specializes`), so if you want to keep the skeleton of your proof the suggestion I'd make is to use the fact that `{0}ᶜ` is open rather than the two open sets. This is a clear case where forcing yourself to use minimal assumptions (T1 instead of T2) leads you to a cleaner proof. Apart from that, the main difference between our proofs is that I know how to avoid being annoyed by `finset.sup` :sweat_smile: